kernel/mellanox_ethernet/src/command_queue.rs

```diff
@@ -0,0 +1,740 @@
+                    opcode, 
+                    opmod, 
+                    None, 
+                    Some(pages_pa.len() as u32)
+                );
+
+                self.init_manage_pages_input_mailbox_buffers(entry_num, pages_pa)?;
+                self.set_mailbox_pointer_in_cmd_entry(&mut cmdq_entry, entry_num, MailboxType::Input);
+            }
+            CommandOpcode::AllocUar => {
+                cmdq_entry.set_input_length_in_bytes(8);
+                cmdq_entry.set_output_length_in_bytes(12);
+                cmdq_entry.set_input_inline_data(opcode, opmod, None, None);
+            },
+            CommandOpcode::QueryVportState => { // only accesses your own vport
+                cmdq_entry.set_input_length_in_bytes(12);
+                cmdq_entry.set_output_length_in_bytes(16);
+                cmdq_entry.set_input_inline_data(opcode, opmod, None, None);
+            },
+            CommandOpcode::AllocPd => { 
+                cmdq_entry.set_input_length_in_bytes(8);
+                cmdq_entry.set_output_length_in_bytes(12);
+                cmdq_entry.set_input_inline_data(opcode, opmod, None, None);
+            },
+            CommandOpcode::AllocTransportDomain => { 
+                cmdq_entry.set_input_length_in_bytes(8);
+                cmdq_entry.set_output_length_in_bytes(12);
+                cmdq_entry.set_input_inline_data(opcode, opmod, None, None);
+            },
+            CommandOpcode::QuerySpecialContexts => { 
+                cmdq_entry.set_input_length_in_bytes(8);
+                cmdq_entry.set_output_length_in_bytes(12);
+                cmdq_entry.set_input_inline_data(opcode, opmod, None, None);
+            },
+            _=> {
+                error!("unimplemented opcode");
+                return Err("Unimplemented opcode!");
+            }
+        }        
+
+        // write command to actual place in memory
+        core::mem::swap(&mut cmdq_entry, &mut self.entries[entry_num]);
+        
+        // update token to be used in the next command
+        self.token = self.token.wrapping_add(1);
+
+        // claim the command entry as in use
+        self.available_entries[entry_num] = false;
+
+        Ok(entry_num)
+    }
+
+    /// Sets the mailbox pointer in a command entry with the physical address of the first mailbox.
+    fn set_mailbox_pointer_in_cmd_entry(&mut self, cmdq_entry: &mut CommandQueueEntry, entry_num: usize, mailbox_type: MailboxType) {
+        let mailbox_ptr = if mailbox_type == MailboxType::Input {
+            self.mailbox_buffers_input[entry_num][0].1.value()
+        } else {
+            self.mailbox_buffers_output[entry_num][0].1.value()
+        };
+        cmdq_entry.input_mailbox_pointer_h.write(U32::new((mailbox_ptr >> 32) as u32));
+        cmdq_entry.input_mailbox_pointer_l.write(U32::new((mailbox_ptr & 0xFFFF_FFFF) as u32));
+    }
+
+    /// Initialize output mailboxes for the QUERY_ISSI command.
```

Comment:
anywhere the comments mention a command, I think it's best to link to the Rust enum/struct version of that rather than plaintext (which i assume is styled according to the mellanox manual). 

Consider that most people reading these docs won't have the mellanox manual in front of them